### PR TITLE
Refactor travis tests and add firefox-beta tests

### LINKED ISSUE #1184: Fix super_cookie_test.py.SuperCookieTests.test_should_detect_ls_of_third_party_frame
[This test](https://github.com/EFForg/privacybadger/blob/master/tests/selenium/super_cookie_test.py#L20) fails intermittently as noted by the docstring. The docstring suggests that the cookie is being set before the pb content script runs. To test this I was tried adding a sleep to the javascript that sets the cookie. My replacement test is [here]( https://rawgit.com/cowlicks/e04533cacf5717cd2cae51e0452032c0/raw/548c8ba38a7c9c67bcb267c0c83a7eccd7b849bf/frame_ls.html). And the gists are [here](https://gist.github.com/cowlicks/e04533cacf5717cd2cae51e0452032c0) and [here](https://gist.github.com/cowlicks/7c6009a982ef3e8cf582b75e5bffb275). This wasn't working when I sidetracked it, but I don't remember why.

### LINKED ISSUE #1267: Add Firefox Beta channel to automated tests
Following up on https://github.com/EFForg/privacybadger/issues/1009#issuecomment-275499800.

### LINKED ISSUE #1294: Stop using mozilla's web-ext to install privacy badger in the selenium tests.
Currently in the selenium tests use [`web-ext`](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Getting_started_with_web-ext) to launch firefox with privacy badger installed. We do this because selenium currently does not support launching FF with WebExtension's installed. Once selenium supports this, we should use it. Then we can remove the `selenium_firefox.sh` script. Part of this is being tracked [here](https://github.com/mozilla/geckodriver/issues/211).

### LINKED ISSUE #1345: Don't install everything for every job in Travis
We shouldn't need to install Firefox (and eventually [beta Firefox](https://github.com/EFForg/privacybadger/issues/1267)) and every Chrome release channel all at the same time for every Travis job. Let's only install the Firefox or Chrome channel we actually use for the given job.

Same for everything else: No need to install both chromedriver and geckodriver, Chrome jobs don't need `web-ext` from package.json, ...

This should result in faster, maybe more reliable (less chance of network failure) Travis builds.

### PR DESCRIPTION
**Pull request by cowlicks**
Fixes #1345, fixes #1184, fixes #1267, fixes #1294.

This might address #1331. I'm going to leave the issue open, and wait for it to re-occur. 

Also partially addresses #1195 in that it adds more thorough documentation for contributors.

### CODE REVIEW

**Comment by ghostwords** [comment]
Would it be possible to label Firefox jobs so that you can tell at a glance which Firefox the job is for, same as you can tell with Chrome jobs (Chrome "stable" vs. "beta")?

**Comment by cowlicks** [comment]
@ghostwords yep, I'll do that

**Comment by ghostwords** [comment]
To answer my own wondering of why we run one set of tests on Python 3: https://github.com/EFForg/privacybadger/pull/933#issuecomment-248549316. I'm not sure we need to support both Python 2 and 3 though ... We could just specify Python 3 (in code and docs) for running all tests and remove the extra job.

**Comment by cowlicks** [comment]
@ghostwords There is a dependency that the firefox selenium tests need that isn't supported in python3. We can do that once selenium has better support for webextensions on firefox.

**Comment by cowlicks** [comment]
I had switched to loading the extension from source for chrome, instead of building CRX file. This was working locally because I was getting a consistent extension URL locally, but on travis it is different everytime. So I've switched back to packing the crx.

**Comment by cowlicks** [comment]
##Some updates,

I've removed the depency on `marionette_driver`, which didn't support python 3, and was only nedded for firefox. This means you get all the dependencies for running the tests with one `pip install -r requirements.txt`. You should still install the `Xvfb` os package if you want to use it.

I've cleaned up the way the tests can be run. Now you just run `BROWSER=... py.test path/to/tests`, where `BROWSER` can be any of path to a browser binary, the name of a browser executable, or just `chrome` or `firefox`.

The tests now run for more browsers.

The scripts for running travis are a lot simpler.

I've tried to clean up some of the tests themselves, but I'm going to remove this and save it for a separate PR.

## problems with firefox:

I've been iterating on how to install firefox, this has been too time consuming. Previously we were launching ff with `web-ext` connecting to it with mozilla's marionette driver to get the extension's URL, then killing marionette and connecting with selenium, for every test.

I tried installing the extension in a firefox profile manually before starting firefox. The problem with this was that it only works on "unbranded" versions of firefox, because these versions don't check for package signatures. Most dev's probably won't have an unbranded version installed, so I dropped this method (but unbranded builds are available on travis).

Currently I use a [new feature](https://github.com/mozilla/geckodriver/releases/tag/v0.17.0) of gecko driver to do a 'temporary install' of the extension. Temporary installs don't require a signature check. Code [here](https://github.com/EFForg/privacybadger/pull/1437/commits/690862e88e22f3c992ca29f87e5264223bcbc6b1#diff-f448424892a6d1a088de9a3cc2c800abR137).

The problem now is that this performs horribly. Locally: firefox frequently coredumps then the systemd journaling service on my machine writes a bunch of crap to my logs, blocking all other I/O for like 10 minutes. Then some other thing crashes from all these writes, and everything just sucks, and some other thing called `[BrowserBlocking]` also likes to crash. This could just be my configuration though. I'd like to better understand this issue so I know if it will effect other people, or if it is just from my particular configuration. I'm also seeing failures on travis and I'm not sure if this is from the poor performance, or just something I changed in the test code.

**Comment by cowlicks** [comment]
The only remaining broken build is `firefox-beta`. Ready for review.

File: tests/selenium/pbtest.py [hunk 1]
```diff
@@ -65,6 +57,24 @@ def get_browser_name(string):
...
+    '''Builds the .crx file for Chrome and returns the path to it'''
+    cmd = ['make', '-sC', get_git_root(), 'travisbuild']
+    return os.path.join(get_git_root(), run_shell_command(cmd).split()[-1])
+
+
+def install_ext_on_ff(driver, extension_path):
+    '''
+    Use Selenium's internal API's to manually send a message to geckodriver
+    to install the extension. We should remove this once the functionality is
+    included in Selenium.
```
Comment (ghostwords) [inline tests/selenium/pbtest.py]: Are there GitHub issues we can reference here?

File: src/js/contentscripts/socialwidgets.js [hunk 2]
```diff
@@ -330,7 +330,7 @@ function replaceIndividualButton(tracker) {
 *                            replaced
 */
 function getTrackerData(callback) {
-  chrome.runtime.sendMessage({checkReplaceButton:document.location}, function(response) {
+  chrome.runtime.sendMessage({checkReplaceButton:document.location.hostname}, function(response) {
```
Comment (ghostwords) [inline src/js/contentscripts/socialwidgets.js]: Should we file a Bugzilla issue for this? New and different behavior from Chrome. Do we need to do a patch release to fix this here (and elsewhere in the code) before Firefox beta lands to stable?

File: tests/selenium/pbtest.py [hunk 3]
```diff
@@ -63,6 +63,18 @@ def build_crx():
     return os.path.join(get_git_root(), run_shell_command(cmd).split()[-1])
 
 
+def install_ext_on_ff(driver, extension_path):
+    '''
+    Use Selenium's internal API's to manually send a message to geckodriver
+    to install the extension. We should remove this once the functionality is
+    included in Selenium.
+    '''
```
Comment (ghostwords) [inline tests/selenium/pbtest.py]: Are there GitHub issues we can reference here?

File: tests/selenium/pbtest.py [hunk 4]
```diff
@@ -1,159 +1,257 @@
...
+    cmd = ['make', '-sC', get_git_root(), 'travisbuild']
+    return os.path.join(get_git_root(), run_shell_command(cmd).split()[-1])
 
 
-def repeat_if_failed(ntimes):
+def install_ext_on_ff(driver, extension_path):
     '''
-    Retry the tests `ntimes` until it passes. Must be used on a subclass of
-    unittest.TestCase
+    Use Selenium's internal API's to manually send a message to geckodriver
```
Comment (ghostwords) [inline tests/selenium/pbtest.py]: Are there GitHub issues we can reference here?

Comment (cowlicks) [inline tests/selenium/pbtest.py]: done

**Comment by cowlicks** [comment]
Test are passing. This is ready for review.

File: src/js/contentscripts/socialwidgets.js (hunk 2 above)
Comment (ghostwords) [inline src/js/contentscripts/socialwidgets.js]: According to https://wiki.mozilla.org/RapidRelease/Calendar, Firefox 55 (current Beta) lands on August 8th, so we need to have a release before then (or Firefox needs to revert their breaking change).

**Comment by ghostwords** [comment]
Re "ghostwords deleted a comment from cowlicks", I was just trying to get rid of the no-longer relevant request to document the relevant Selenium issue.

File: README.md [hunk 5]
```diff
@@ -32,11 +32,37 @@ We also have a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadg
...
 along with [Travis CI](https://travis-ci.org/) for continuous integration.
 
-To run unit tests, find your extension's ID (look for it on `chrome://extensions/`) and
-visit `chrome-extension://YOUR_EXTENSION_ID/tests/index.html`, replacing
-`YOUR_EXTENSION_ID` with your 32 character ID.
-
-To run functional tests powered by Selenium, you need to have `chromedriver`, `xvfb` and `python-virtualenv` installed. Also `geckodriver` to run functional tests in Firefox. See [scripts/setup_travis.sh](scripts/setup_travis.sh) and [tests/run_selenium_tests.sh](tests/run_selenium_tests.sh).
+#### Unit tests
+
+To run unit tests on Chrome find your extension's ID look for it on
```
Comment (ghostwords) [inline README.md]: Nit: Missing "(" between "ID" and "look".

File: README.md [hunk 6]
```diff
@@ -32,11 +32,37 @@ We also have a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadg
...
+`chrome://extensions/`) and visit
+`chrome-extension://YOUR_EXTENSION_ID/tests/index.html`, replacing
+`YOUR_EXTENSION_ID` with your 32 character ID. For Firefox you can find your
+extensions's "UUID" on `about:debugging`, then visit
```
Comment (ghostwords) [inline README.md]: I don't see any IDs on `about:debugging`:

screenshot from 2017-06-23 12 05 20

File: README.md [hunk 7]
```diff
@@ -32,11 +32,37 @@ We also have a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadg
...
+`moz_extension://YOUR_EXTENSION_UUID/tests/index.html`.
+
+#### Functional tests
+
+To run the Selenium functional tests, you'll need to install `chromedriver` for
+Chrome or `geckodriver` for Firefox. Optionally you can also install `xvfb` to
```
Comment (ghostwords) [inline README.md]: Do we need to call out `xvfb`? It gets installed by running `pip` below.

File: .eslintignore [hunk 8]
```diff
@@ -1,3 +1,4 @@
 src/lib/
 src/tests/lib/vendor/
 node_modules/
+privacy_badger*
```
Comment (ghostwords) [inline .eslintignore]: Is this still needed?

**Comment by ghostwords** [comment]
Do the tests not work with Python 3.4.3? That's the Python 3 version on Ubuntu 14.04 LTS, probably other platforms. All tests fail with `SyntaxError: can use starred expression only as assignment target`. Can we work around this incompatibility?

File: README.md [hunk 9]
```diff
@@ -32,11 +32,37 @@ We also have a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadg
...
+run the tests headlessly. You also need some python packages which can be installed by running:
+```bash
+$ pip install -r tests/requirements.txt
+```
+
+Now you should be able to run the selenium tests! Try running
+```bash
+$ BROWSER=chrome py.test
```
Comment (ghostwords) [inline README.md]: We should have an example of how to run only one test.

**Comment by ghostwords** [comment]
One of the tests, `tests/selenium/super_cookie_test.py::SuperCookieTest::test_should_detect_ls_of_third_party_frame`, used to consistently show up in the "slowest 10 test durations" list, but no longer does. Should check the test still does meaningful work.

File: README.md (hunk 9 above)
Comment (ghostwords) [inline README.md]: And how to run without browser windows popping up. Maybe should instead default to headless and show a way to turn that off.

File: README.md [hunk 10]
```diff
@@ -32,11 +32,37 @@ We also have a [mailing list](https://lists.eff.org/mailman/listinfo/privacybadg
...
+```
+
+The BROWSER environment variable must be set. It must be one of:
+* `BROWSER=/path/to/a/browser`
+* the name of a browser executable that can be found like `which $BROWSER`
+* or simply `BROWSER=chrome` or `BROWSER=firefox` if you have them installed
+
+Refer to the our Travis-CI scripts for more information:
+[`scripts/setup_travis.sh`](scripts/setup_travis.sh) and
+[`tests/run_selenium_tests.sh`](tests/run_selenium_tests.sh).
```
Comment (ghostwords) [inline README.md]: Should this be `scripts/run_travis.sh`?

**Comment by ghostwords** [comment]
Could you update the description of this pr to call out the [`tests`-labeled issues](https://github.com/EFForg/privacybadger/issues?q=is%3Aissue+is%3Aopen+label%3Atests) this resolves or at least partially addresses?

File: src/js/popup.js [hunk 11]
```diff
@@ -277,7 +277,7 @@ function registerToggleHandlers() {
 *
 * @param {Integer} tabId The id of the tab
 */
-function refreshPopup(tabId) {
+function refreshPopup(tabId, callback) {
```
Comment (ghostwords) [inline src/js/popup.js]: This should probably be backed out from this pr.

File: src/js/contentscripts/socialwidgets.js (hunk 2 above)
Comment (ghostwords) [inline src/js/contentscripts/socialwidgets.js]: If we are not using `checkReplaceButton`'s value back in `webrequest.js`, let's send `true` instead.

File: tests/selenium/cookie_test.py [hunk 12]
```diff
@@ -206,21 +201,15 @@ def get_tracker_state(self):
         """Parse the UI to group all third party origins into their respective action states."""
 
         # give asynchronously-rendered tracker list time to load
```
Comment (ghostwords) [inline tests/selenium/cookie_test.py]: Comment should be moved to above the `WebDriverWait` line (right?).

File: README.md (hunk 7 above)
Comment (cowlicks) [inline README.md]: Pip only installs a python api that wraps xvfb. So it still need to be installed manually.

File: README.md (hunk 10 above)
Comment (cowlicks) [inline README.md]: Whoops, yep.

File: src/js/contentscripts/socialwidgets.js (hunk 2 above)
Comment (cowlicks) [inline src/js/contentscripts/socialwidgets.js]: Noted. I'll make a bugzilla issue this weekend or monday.

Comment (ghostwords) [inline src/js/contentscripts/socialwidgets.js]: Cool. Bugzilla is kind of a pain. For 
"product" you'll probably want "Toolkit" and for "component" whichever "WebExtensions" component seems to fit best.

**Comment by cowlicks** [comment]
@ghostwords That test had a long (5 second) sleep before, it also was a `repeat_if_failed` test.

The problem was that it had a racecondition. Privacy badger would inject a content script to look for high entropy values in local storage, and the test webpage would write to localstorage. If privacy badger checked before the page was written to, the test would fail.
To circumvent this problem, I load the page, then refresh it, then check that PB detected tracking. Reverting this change breaks firefox-beta tests. I think further changes to the test, or to the content script should be done in a separate PR.

File: tests/selenium/cookie_test.py (hunk 12 above)
Comment (cowlicks) [inline tests/selenium/cookie_test.py]: moved

File: src/js/popup.js (hunk 11 above)
Comment (cowlicks) [inline src/js/popup.js]: removed

File: .eslintignore (hunk 8 above)
Comment (cowlicks) [inline .eslintignore]: removed

File: README.md (hunk 9 above)
Comment (cowlicks) [inline README.md]: Added that stuff below. Mac users can't use Xvfb, so I default to having it disabled unless `ENABLE_XVFB=1` is provided.

**Comment by cowlicks** [comment]
@ghostwords added list of related issues

**Comment by cowlicks** [comment]
Switched to python 3.4

**Comment by ghostwords** [comment]
I edited the description: auto-closing doesn't parse commas/spaces, plus we don't want to auto-close the latter two issues (replaced "resolve" with "address").

**Comment by ghostwords** [comment]
I think we still have an issue with the Python version: Travis, even though it claims it uses "trusty", provides Python 3.4.5 (when told "3.4") but actual "trusty" repos provide Python 3.4.3, which is why the builds aren't running into "SyntaxError: can use starred expression only as assignment target" yet.

**Comment by ghostwords** [comment]
Should we break out testing instructions into own doc and link to it from the README?

**Comment by cowlicks** [comment]
Just for posterity, @ghostwords and I discussed his comments off line. The tests work now for all python 3. And the test documentation is large, but it will be consolidated in a seperate pull request once some other documentation pull requests are merged.

File: src/js/contentscripts/socialwidgets.js (hunk 2 above)
Comment (ghostwords) [inline src/js/contentscripts/socialwidgets.js]: Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1376969.

**Comment by ghostwords** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1376969 for the document.location message passing issue in Firefox Beta (55).

File: src/js/contentscripts/socialwidgets.js (hunk 2 above)
Comment (cowlicks) [inline src/js/contentscripts/socialwidgets.js]: @ghostwords thank you I forgot about this.

**Pull request merged**